src/engine/bank08.asm

```diff
@@ -6219,8 +6222,1319 @@ AIDecide_PokemonTrader_Flamethrower: ; 22133 (8:6133)
+
+.next_1
+	inc e
+	ld a, e
+	cp d
+	jr nz, .loop_play_area_1 ; reached end of Play Area
+
+	ld a, 1
+	cp b
+	jr nc, .failed ; return if less than 2 cards with damage
+
+; card in Play Area with lowest HP remaining was found.
+; look for another card to take damage counter from.
+	ld a, h
+	ldh [hTempRetreatCostCards], a
+	ld b, a
+	ld a, 10
+	cp c
+	jr z, .hp_10_remaining
+	; if has more than 10 HP remaining,
+	; skip Arena card in choosing which
+	; card to take damage counter from.
+	ld e, PLAY_AREA_BENCH_1
+	jr .second_card
+
+.hp_10_remaining
+	; if Curse can KO, then include
+	; Player's Arena card to take
+	; damage counter from.
+	ld e, PLAY_AREA_ARENA
+
+.second_card
+	ld a, DUELVARS_NUMBER_OF_POKEMON_IN_PLAY_AREA
+	call GetTurnDuelistVariable
+	ld d, a
+.loop_play_area_2
+	ld a, e
+	cp b
+	jr z, .next_2 ; skip same Pokemon card
+	push bc
+	call GetCardDamage
+	pop bc
+	jr nz, .use_curse ; has damage counters, choose this card
+.next_2
+	inc e
+	ld a, e
+	cp d
+	jr nz, .loop_play_area_2
+
+.failed
+	call SwapTurn
+	or a
+	ret
+
+.use_curse
+	ld a, e
+	ldh [hAIPkmnPowerEffectParam], a
+	call SwapTurn
+	ld a, [wce08]
+	ldh [hTempCardIndex_ff9f], a
+	ld a, OPPACTION_USE_PKMN_POWER
+	bank1call AIMakeDecision
+	ld a, OPPACTION_EXECUTE_PKMN_POWER_EFFECT
+	bank1call AIMakeDecision
+	ld a, OPPACTION_DUEL_MAIN_SCENE
+	bank1call AIMakeDecision
+	ret
+; 0x2262d
+
+; handles AI logic for Cowardice
+HandleAICowardice: ; 2262d (8:662d)
+	ld a, MUK
+	call CountPokemonIDInBothPlayAreas
+	ret c ; return if there's Muk in play
+
+	farcall AIChooseRandomlyNotToDoAction
+	ret c ; randomly return
+
+	ld a, DUELVARS_NUMBER_OF_POKEMON_IN_PLAY_AREA
+	call GetTurnDuelistVariable
+	cp 1
+	ret z ; return if only one Pokemon in Play Area
+
+	ld b, a
+	ld c, PLAY_AREA_ARENA
+	ld a, DUELVARS_ARENA_CARD_STATUS
+	call GetTurnDuelistVariable
+	and CNF_SLP_PRZ
+	jr nz, .next
+.loop
+	ld a, DUELVARS_ARENA_CARD
+	add c
+	call GetTurnDuelistVariable
+	ld [wce08], a
+	call GetCardIDFromDeckIndex
+	ld a, e
+	push bc
+	cp TENTACOOL
+	call z, .CheckWhetherToUseCowardice
+	pop bc
+	jr nc, .next
+
+	dec b ; subtract 1 from number of Pokemon in Play Area
+	ld a, 1
+	cp b
+	ret z ; return if no longer has Bench Pokemon
+	ld c, PLAY_AREA_ARENA ; reset back to Arena
+	jr .loop
+
+.next
+	inc c
+	ld a, c
+	cp b
+	jr nz, .loop
+	ret
+; 0x22671
+
+; checks whether AI uses Cowardice.
+; return carry if Pkmn Power was used.
+; input:
+;	c = Play Area location (PLAY_AREA_*) of Tentacool.
+.CheckWhetherToUseCowardice ; 22671 (8:6671)
+	ld a, c
+	ldh [hTemp_ffa0], a
+	ld e, a
+	call GetCardDamage
+.asm_22678
+	or a
+	ret z ; return if has no damage counters
+
+	ldh a, [hTemp_ffa0]
+	or a
+	jr nz, .is_benched
+	farcall AIDecideBenchPokemonToSwitchTo
+; in case this routine is called with no Bench Pokemon,
+; this introduces a bug.
+; since it's only called in HandleAICowardice in case
+; there's more than 1 Pokemon in Play Area, this never sets carry.
```

Comment:
I'm not really sure what you're saying the bug is

Reply:
It's not a bug simply because the routine is protected against it. Basically, the intention was to return with no carry from this routine if AIDecideBenchPokemonToSwitchTo returns carry (which means there are no Bench Pokemon). But if that happened, the jr c command would erroneously jump to the middle of the damage check above. And a can have any value at this point, so it would _not_ return and instead loop back around, potentially making an infinite loop. The point is that this never happens because AIDecideBenchPokemonToSwitchTo was already called before and will never return carry here.

I guess I wasn't explicit on what "this introduces a bug" was referring to, basically the very next 2 lines should be something like jr nc, .use_cowardice, xor a, or a, ret. I'll reword this part.

Reply:
Actually, the jr c can just be safely removed since it will never return carry. That's way simpler.